### PR TITLE
feat: consolidate Slack reports into one message per cycle with collapsible project sections

### LINKED ISSUE #179: feat: consolidate Slack reports into one message per cycle with collapsible project sections
## Problem

Currently oompa sends one Slack message **per project per poll cycle**. With 8 projects, that's potentially 8 separate Slack messages every 2 minutes — noisy and hard to scan.

## Desired Behavior

Send a **single Slack message per poll cycle** across all projects, with a header and collapsible project sections. At a glance you see the report header and project names. Click a project to expand its findings.

### Slack message format

\`\`\`
🏭 oompa report — 3 projects with activity

▸ ovn-kubernetes/ovn-kubernetes (2 PRs)
  📋 PR #6252 — Fix kubevirt live migration test flake
    🔴 e2e (shard-conformance) — failed
    🔴 test-deploy — infra (GitHub 500)
    ⚠️ 15 commits behind main
  📋 PR #6118 — kubevirt: Support vm hostname
    💬 3 new reviews from coderabbitai
    ⚠️ Merge conflicts

▸ openshift/hypershift (1 PR)
  📋 PR #8365 — KubeVirt nmstate conditional
    🔴 e2e-kubevirt-aws-ovn — 15 test failures
    ✅ Rebased → def5678

▸ qinqon/oompa (1 issue)
  🔧 Issue #177 — PR #178 created
    ✅ CI passing
    💬 2 reviews from gemini-code-assist
\`\`\`

### When collapsed (default view in Slack)

The user sees just the header and project lines:

\`\`\`
🏭 oompa report — 3 projects with activity

▸ ovn-kubernetes/ovn-kubernetes (2 PRs)
▸ openshift/hypershift (1 PR)
▸ qinqon/oompa (1 issue)
\`\`\`

Clicking a project expands to show its PR/issue details.

### Implementation with Slack Block Kit

Use Slack's \`rich_text\` blocks with \`rich_text_list\` for the collapsible structure, or use the simpler approach of \`section\` blocks with \`mrkdwn\`:

\`\`\`json
{
  "blocks": [
    {
      "type": "header",
      "text": {
        "type": "plain_text",
        "text": "🏭 oompa report — 3 projects with activity"
      }
    },
    {
      "type": "section",
      "text": {
        "type": "mrkdwn",
        "text": "*<https://github.com/ovn-kubernetes/ovn-kubernetes|ovn-kubernetes/ovn-kubernetes>* (2 PRs)"
      }
    },
    {
      "type": "section",
      "text": {
        "type": "mrkdwn",
        "text": "📋 <https://github.com/ovn-k/ovn-k/pull/6252|PR #6252> — Fix kubevirt test flake\n  🔴 <https://job-link|e2e (shard-conformance)> — failed\n  ⚠️ 15 commits behind main\n📋 <https://github.com/ovn-k/ovn-k/pull/6118|PR #6118> — kubevirt vm hostname\n  💬 3 new reviews from <https://github.com/coderabbitai|coderabbitai>"
      }
    },
    {"type": "divider"},
    ...next project...
  ]
}
\`\`\`

Note: Slack doesn't have native collapsible/accordion sections in Block Kit. The closest approaches are:
1. **Separate blocks per project** with the project name as a bold header — users can visually scan project names and read details below each one
2. **Slack's \`<details>\` in mrkdwn** — not supported, but the visual structure with indentation and dividers achieves a similar effect
3. **Attachment-based collapsing** — Slack collapses long attachments with a "Show more" link. If each project's content is in a separate attachment, Slack may auto-collapse verbose ones

### Key changes from current behavior

| Current | Proposed |
|---------|----------|
| One message per project per cycle | One message per cycle across all projects |
| Each goroutine posts independently | Findings collected centrally, posted once at end of cycle |
| Repeated messages if same finding persists | Dedup: report once, suppress until change |

### Collection architecture

Currently each \`Agent\` has its own \`SlackReporter\` and posts independently. The change requires:

1. **Centralized finding collector** — a shared \`SlackFindings\` channel or mutex-protected slice that all goroutines append to
2. **Cycle-end aggregator** — after all goroutines complete their poll cycle (or on a timer), collect all findings, group by project, format as one message, POST once
3. **One webhook call per cycle** — reduces API calls from N (one per project) to 1

### Timing consideration

Since goroutines run independently and may finish at different times, the aggregator needs to either:
- Wait for all goroutines to complete their cycle (complex synchronization)
- Use a **time-based window** — collect findings over a 2-minute window, then flush (simpler, matches poll interval)
- Post at the **start of each cycle** with findings from the previous cycle (simplest)

### Projects with no findings

Projects where nothing happened are **not included** in the message at all. Only projects with actual findings appear. If no project has findings, no Slack message is sent.

### Links

Every item must link to the relevant GitHub resource:
- Project name → \`github.com/owner/repo\`
- PR number → \`github.com/owner/repo/pull/N\`
- CI check name → \`CheckRun.HTMLURL\`
- Flaky issue → \`github.com/owner/repo/issues/N\`
- Commit SHA → \`github.com/owner/repo/pull/N/commits/SHA\`
- Reviewer → \`github.com/username\`

## Files to change

| File | Change |
|------|--------|
| \`pkg/agent/slack.go\` | Refactor from per-agent posting to centralized collector + single message formatter |
| \`pkg/agent/loop.go\` | Collect findings into shared structure instead of posting per-agent |
| \`cmd/oompa/main.go\` | Create shared finding collector, pass to agents, flush after cycle |

## Tests

- Multiple projects with findings → single Slack message with all projects
- One project with findings, others idle → message only shows that project
- No findings anywhere → no Slack message
- Links are correct for PRs, CI jobs, issues, commits
- Message stays under Slack's 3000-char limit per block (split into multiple blocks if needed)
- Dedup: same finding on consecutive cycles → only first sends

### PR DESCRIPTION
**Pull request by qinqon**
Fixes #179

## Summary

Consolidates Slack reports into a single message per poll cycle across all projects, replacing the previous per-project-per-cycle approach that could produce N separate Slack messages (one per project goroutine).

## Changes

- Refactored `SlackReporter` to be project-agnostic with thread-safe `Collect()` and `Flush()` methods instead of per-project `Report()` calls
- Added `Owner`/`Repo` fields to `SlackFinding` so each finding carries its project identity
- Rewrote `formatSlackMessage` to group findings by project first, then by PR, using Slack Block Kit with header, per-project section headers with repo links, detail sections, and dividers
- Added block text splitting to stay within Slack's 3000-character limit per block
- In multi-project mode: created a shared `SlackReporter` passed to all agent goroutines, with a separate flush goroutine that batches findings on a timer matching the poll interval
- In single-repo mode: findings are collected and flushed at the end of each poll cycle
- Updated the Slack spec to document the new consolidated architecture
- Added new tests: multi-project formatting, concurrent collection, collect/flush lifecycle, block splitting, divider placement, link correctness

## Testing

- [x] `go test ./...` passes
- [x] `go vet ./...` passes
- [ ] Manual testing (if applicable)

## Related Issues

Fixes #179

## Summary by CodeRabbit

* **New Features**
  * Consolidated per-poll-cycle Slack reports across projects with per-project sections, clickable GitHub links, overflow splitting and truncation; shared, thread-safe reporter with periodic/background flush; single-repo convenience path retained.
* **Bug Fixes**
  * Deduplication across cycles to avoid duplicate notifications; silent/idle cycles produce no message; final flush on shutdown; thread-safe collection to prevent concurrent flush races.
* **Documentation**
  * Updated Slack integration spec covering Block Kit format, dedup, collect/flush and truncation behavior.
* **Tests**
  * Expanded tests for formatting, splitting/truncation, dedup, concurrency, disabled mode, and collect+flush pipeline.

[Review Change Stack](https://app.coderabbit.ai/change-stack/qinqon/oompa/pull/180)

### CODE REVIEW

File: pkg/agent/slack.go [hunk 1]
```diff
@@ -129,68 +169,156 @@ type slackMessage struct {
...
+		// Split into multiple blocks if exceeding Slack's 3000-char limit
+		for len(detailText) > 0 {
+			chunk := detailText
+			if len(chunk) > maxSlackBlockTextLen {
+				// Find the last newline within the limit to avoid cutting mid-line
+				cutIdx := strings.LastIndex(chunk[:maxSlackBlockTextLen], "\n")
+				if cutIdx <= 0 {
+					cutIdx = maxSlackBlockTextLen
+				}
+				chunk = detailText[:cutIdx+1]
```
[the remark this answers is not in the document]
Comment (qinqon) [reply pkg/agent/slack.go]: > The block-splitting logic can generate a chunk longer than `maxSlackBlockTextLen`

Addressed: Fixed the off-by-one error. Changed `cutIdx <= 0` to `cutIdx < 0` (since `strings.LastIndex` returns -1 on not-found, not 0), and restructured the fallback to use `detailText[:maxSlackBlockTextLen]` directly when no newline is found, ensuring chunks never exceed the limit. Added `TestFormatSlackMessage_BlockTextLimit_SingleLongLine` to cover the no-newline edge case.

File: cmd/oompa/main.go [hunk 2]
```diff
@@ -726,6 +761,10 @@ func runMultiProject(globalCfg agent.Config, configPath string, ghClient *agent.
 	}
 
 	wg.Wait()
+	// Final flush to capture any findings from the last cycle
+	if sharedSlack != nil {
+		sharedSlack.Flush(context.Background())
+	}
```
[the remark this answers is not in the document]
Comment (qinqon) [reply cmd/oompa/main.go:775]: > the periodic Slack flush goroutine is only stopped when `ctx` is canceled. If all role goroutines exit without canceling the shared context...the final `sharedSlack.Flush(...)` below can run concurrently with a ticker flush.

Addressed two ways:

1. **Structural fix**: The flush goroutine now signals completion via a `flushDone` channel. After `wg.Wait()`, we wait for `<-flushDone` before calling the final `Flush()`, ensuring no concurrent Flush calls.

2. **Defense in depth**: Added `flushMu sync.Mutex` to `SlackReporter` that serializes all `Flush()` calls, making it safe even if called concurrently.

File: pkg/agent/slack_test.go [hunk 3]
```diff
@@ -427,3 +586,158 @@ func TestURLHelpers(t *testing.T) {
...
+		t.Fatalf("invalid JSON: %v", err)
+	}
+
+	// Verify no block text exceeds the limit
+	for i, b := range msg.Blocks {
+		if b.Text != nil && len(b.Text.Text) > maxSlackBlockTextLen {
+			t.Errorf("block %d text exceeds limit: %d > %d", i, len(b.Text.Text), maxSlackBlockTextLen)
+		}
+	}
+}
```
[the remark this answers is not in the document]
Comment (qinqon) [reply pkg/agent/slack_test.go:743]: > `TestFormatSlackMessage_BlockTextLimit` only covers splitting when there are many newline-separated lines...Add a test with one extremely long PR title/message line

Addressed: Added `TestFormatSlackMessage_BlockTextLimit_SingleLongLine` that creates a single finding with a message exceeding 3000 chars containing no newlines. The test verifies no block exceeds the limit and that at least 2 detail blocks are produced from the split.

File: specs/slack.md [hunk 4]
```diff
@@ -12,16 +14,18 @@ Automatic Slack reporting at the end of each poll cycle. When `OOMPA_SLACK_WEBHO
...
 |------|---------|
-| `pkg/agent/slack.go` | SlackFinding, SlackReporter, postToSlack, message formatter, dedup tracker |
-| `pkg/agent/slack_test.go` | Tests for formatting, dedup, empty findings, link construction |
-| `pkg/agent/loop.go` | Agent loop integration: Check* methods, ShouldCheckReaction, RunReportOnlyChecks, cycle-end reporting |
+| `pkg/agent/slack.go` | SlackFinding, SlackReporter (thread-safe collector + flusher), postToSlack, message formatter, dedup tracker |
+| `pkg/agent/slack_test.go` | Tests for formatting, dedup, empty findings, link construction, collect/flush, concurrency, block splitting |
+| `pkg/agent/loop.go` | Agent loop integration: Check* methods, ShouldCheckReaction, RunReportOnlyChecks, CollectSlackFindings, FlushSlackReport |
 | `pkg/agent/config.go` | `SlackWebhookURL` field on Config |
-| `cmd/oompa/main.go` | Read `OOMPA_SLACK_WEBHOOK` env var |
+| `cmd/oompa/main.go` | Read `OOMPA_SLACK_WEBHOOK` env var, create shared reporter in multi-project mode, flush goroutine |
```
[the remark this answers is not in the document]
Comment (qinqon) [reply specs/slack.md:22]: > The spec claims `SlackReporter` is a "thread-safe collector + flusher", but the current implementation only protects the `pending` buffer

Addressed: Added `flushMu sync.Mutex` to `SlackReporter` that serializes all `Flush()` calls, protecting the `reported` map from concurrent access. Updated both the spec and the code doc to accurately describe the dual-mutex design: `mu` protects the pending buffer, `flushMu` serializes Flush.

File: pkg/agent/slack.go [hunk 5]
```diff
@@ -129,68 +169,156 @@ type slackMessage struct {
...
+			}
```
[the remark this answers is not in the document]
Comment (qinqon) [reply pkg/agent/slack.go:322]: > Block splitting can exceed the configured max by 1 byte in the no-newline fallback.

Addressed: Fixed the off-by-one. Changed the condition to `cutIdx < 0` (matching `strings.LastIndex` semantics) and restructured to use `detailText[:maxSlackBlockTextLen]` directly in the no-newline fallback, exactly as suggested. Added `TestFormatSlackMessage_BlockTextLimit_SingleLongLine` to cover this edge case.

File: pkg/agent/slack.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (qinqon) [reply pkg/agent/slack.go]: > There is an off-by-one error in the block splitting logic.

Addressed: Fixed the off-by-one. Changed `cutIdx <= 0` to `cutIdx < 0` and restructured the fallback to hard-cut at `maxSlackBlockTextLen` when no newline is found. Added test `TestFormatSlackMessage_BlockTextLimit_SingleLongLine` to cover the no-newline edge case.

File: pkg/agent/slack.go [hunk 6]
```diff
@@ -129,68 +169,156 @@ type slackMessage struct {
...
+		// Divider between projects
+		blocks = append(blocks, slackBlock{Type: "divider"})
+	}
+
+	// Remove trailing divider
+	if len(blocks) > 0 && blocks[len(blocks)-1].Type == "divider" {
+		blocks = blocks[:len(blocks)-1]
 	}
 
+	msg := slackMessage{Blocks: blocks}
```
[the remark this answers is not in the document]
Comment (qinqon) [reply pkg/agent/slack.go:355]: > Slack has a limit of 50 blocks per message...Consider batching blocks into multiple messages if the limit is reached.

Addressed: Added `maxSlackBlocks = 50` constant and truncation logic in `formatSlackMessage`. When blocks exceed 50, the message is truncated at 49 blocks and a final overflow notice block is appended. Added `TestFormatSlackMessage_TruncatesAt50Blocks` to verify. Chose truncation with notice over multi-message batching — Slack webhooks lack threading support needed for clean multi-message delivery, and truncation ensures the message is always delivered.

File: pkg/agent/slack_test.go [hunk 7]
```diff
@@ -28,51 +29,197 @@ func TestFormatSlackMessage_GroupsByPR(t *testing.T) {
...
+		t.Error("expected project header to contain org/repo")
+	}
+
+	// Detail section — find the section with PR details
+	fullText := ""
+	for _, b := range msg.Blocks {
+		if b.Text != nil {
+			fullText += b.Text.Text
+		}
+	}
```
[the remark this answers is not in the document]
Comment (qinqon) [reply pkg/agent/slack_test.go]: > Replace the repeated string concatenation when building fullText from msg.Blocks with a strings.Builder

Addressed: Replaced both occurrences of `fullText := "" / fullText += b.Text.Text` with `strings.Builder` in `TestFormatSlackMessage_GroupsByPR` and `TestFormatSlackMessage_MultipleProjects`.

File: pkg/agent/slack_test.go [hunk 8]
```diff
@@ -427,3 +586,158 @@ func TestURLHelpers(t *testing.T) {
...
+
+	logger := slog.New(slog.NewTextHandler(io.Discard, nil))
+	reporter := NewSlackReporter(ts.URL, logger)
+	reporter.httpClient = ts.Client()
+
+	// Simulate concurrent collection from multiple goroutines
+	var wg sync.WaitGroup
+	for i := 0; i < 10; i++ {
+		wg.Add(1)
+		i := i
```
[the remark this answers is not in the document]
Comment (qinqon) [reply pkg/agent/slack_test.go]: > Replace the counted for-loops that use the index with an explicit integer-range form

Addressed: Changed both `for i := 0; i < N; i++` loops to `for i := range N` form, removing the now-unnecessary `i := i` shadow variable.

**Comment by qinqon** [comment]
> `Flush` is not safe under concurrent calls and can race on `reported`.
> — @coderabbitai (outside diff range), @copilot-pull-request-reviewer (suppressed low-confidence)

Addressed: Added `flushMu sync.Mutex` to `SlackReporter` that serializes all `Flush()` calls, protecting the `reported` map from concurrent access. Additionally, the flush goroutine in `main.go` now signals completion via a `flushDone` channel, ensuring the final `Flush()` only runs after the ticker goroutine exits.

Updated the spec to accurately describe the dual-mutex design.

**Comment by qinqon** [comment]
> **Use UTF-8 boundary-safe slicing when truncating long single-line text.** On line 309, `detailText[:maxSlackBlockTextLen]` slices by bytes. For text containing multibyte characters (emoji, CJK text, etc.), this can split mid-rune and produce invalid UTF-8.

Addressed: Added `unicode/utf8` import and backtracking logic in the no-newline fallback path. When no newline is found within the limit, the cut index backs up from `maxSlackBlockTextLen` until `utf8.ValidString(detailText[:cut])` is true, ensuring we never split mid-rune. Also fixed indentation of the block-splitting logic that was misaligned from a previous edit.

**Comment by qinqon** [comment]
Rebased commit 30a3787 on main and pushed.

**Comment by qinqon** [comment]
## CI Failure Analysis — commit 1bfa815

**Total failures**: 1 (1 related)
**Action taken**: Pushed fix for related failure

---

<details>
<summary>🔴 Related: <code>test</code> — fix pushed</summary>

### Error
````
```
test	Check go fix	2026-05-12T14:29:19.4239954Z @@ -652,13 +652,11 @@ func TestSlackReporter_CollectConcurrent(t *testing.T)
test	Check go fix	2026-05-12T14:29:19.4242372Z -		wg.Add(1)
test	Check go fix	2026-05-12T14:29:19.4242978Z -		go func() {
test	Check go fix	2026-05-12T14:29:19.4243377Z -			defer wg.Done()
test	Check go fix	2026-05-12T14:29:19.4243725Z +		wg.Go(func() {
test	Check go fix	2026-05-12T14:29:19.4248503Z go fix ./... produced changes; run 'make fix' and commit
```
````

### Root Cause
Go 1.25 introduced `sync.WaitGroup.Go()` as a replacement for the manual `Add(1)/go func()/Done()` pattern. The CI runs `make check-fix` which runs `go fix ./...` and checks for uncommitted changes. The test code in `slack_test.go` used the old pattern which `go fix` auto-rewrites.

### Recommendation
Fixed by replacing `wg.Add(1); go func() { defer wg.Done(); ... }()` with `wg.Go(func() { ... })` in `TestSlackReporter_CollectConcurrent`.

### Action
Pushed fix for this failure.

</details>

Pushed a fix for the related failure.

File: pkg/agent/slack.go [hunk 9]
```diff
@@ -129,68 +180,175 @@ type slackMessage struct {
...
+			}
 		}
-	}
 
-	msg := slackMessage{
-		Blocks: []slackBlock{
-			{
+		detailText := sb.String()
+		// Split into multiple blocks if exceeding Slack's 3000-char limit
+		for len(detailText) > 0 {
```
[the remark this answers is not in the document]
Comment (qinqon) [reply pkg/agent/slack.go]: > **Fix linter issue: use idiomatic empty string check.**
> Replace `len(detailText) > 0` with `detailText != ""` per gocritic's `emptyStringTest` rule.

Addressed: Replaced `len(detailText) > 0` with `detailText != ""` on line 303. Also added a `cut == 0` guard below to prevent a theoretical infinite loop with malformed UTF-8 input (from a related nitpick in the same review round).

**Comment by qinqon** [comment]
> **Consider adding a guard against empty chunk to prevent theoretical infinite loop.**
> If `cut` reaches 0 (theoretically possible with severely malformed UTF-8), `chunk` becomes empty and the loop never advances.

Addressed: Added `if cut == 0 { break }` guard after the UTF-8 boundary backtracking loop to prevent a theoretical infinite loop with malformed input.

**Comment by qinqon** [comment]
Rebased commit 85baee9 on main and pushed.
